fendermint/vm/topdown/src/vote/mod.rs

```diff
@@ -0,0 +1,130 @@
+use crate::vote::operation::{OperationMetrics, OperationStateMachine};
+use crate::BlockHeight;
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use tokio::sync::{broadcast, mpsc};
+
+#[derive(Clone)]
+pub struct VoteRecord {}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Config {
+    /// The reactor request channel buffer size
+    req_channel_buffer_size: usize,
+    /// The number of requests the reactor should process per run before handling other tasks
+    req_batch_processing_size: usize,
+    /// The number of vote recording requests the reactor should process per run before handling other tasks
+    gossip_req_processing_size: usize,
+    /// The time to sleep for voting loop if nothing happens
+    voting_sleep_interval_sec: u64,
+}
+
+pub struct VoteReactorClient {
+    tx: mpsc::Sender<VoteReactorRequest>,
+}
+
+pub fn start_vote_reactor(
+    config: Config,
+    gossip_rx: broadcast::Receiver<VoteRecord>,
+    gossip_tx: mpsc::Sender<VoteRecord>,
+    internal_event_listener: broadcast::Receiver<TopDownSyncEvent>,
+) -> VoteReactorClient {
+    let (tx, rx) = mpsc::channel(config.req_channel_buffer_size);
+
+    tokio::spawn(async move {
+        let sleep = Duration::new(config.voting_sleep_interval_sec, 0);
+
+        let inner = VotingHandler {
+            req_rx: rx,
+            gossip_rx,
+            gossip_tx,
+            internal_event_listener,
+            config,
+        };
+        let mut machine = OperationStateMachine::new(inner);
+        loop {
+            machine = machine.step();
+            tokio::time::sleep(sleep).await;
+        }
+    });
+
+    VoteReactorClient { tx }
+}
+
+enum VoteReactorRequest {
+    QueryOperationMode,
+    QueryVotes(BlockHeight),
+}
+
+struct VotingHandler {
+    /// Handles the requests targeting the vote reactor, could be querying the
+    /// vote tally status and etc.
+    req_rx: mpsc::Receiver<VoteReactorRequest>,
+    /// Receiver from gossip pub/sub, mostly listening to incoming votes
+    gossip_rx: broadcast::Receiver<VoteRecord>,
+    gossip_tx: mpsc::Sender<VoteRecord>,
+    /// Listens to internal events and handles the events accordingly
+    internal_event_listener: broadcast::Receiver<TopDownSyncEvent>,
+    config: Config,
+}
+
+impl VotingHandler {
+    fn handle_request(&self, _req: VoteReactorRequest) {}
+
+    fn record_vote(&self, _vote: VoteRecord) {}
+
+    fn handle_event(&self, _event: TopDownSyncEvent) {}
+
+    fn process_external_request(&mut self, _metrics: &OperationMetrics) -> usize {
+        let mut n = 0;
+        while n < self.config.req_batch_processing_size {
+            match self.req_rx.try_recv() {
+                Ok(req) => {
+                    self.handle_request(req);
+                    n += 1
+                }
+                Err(mpsc::error::TryRecvError::Disconnected) => {
+                    tracing::warn!("voting reactor tx closed unexpected");
+                    break;
```

Comment:
`break` will only break the batch loop, before re-entering. There's no shutdown / error handling mechanism.